### ISSUE TITLE
Multiple font families not handled in OSX

### ISSUE THREAD

**Issue by carljparker** [Kozea/WeasyPrint#158]
Specifying multiple comma-separated names in `font-family` doesn’t work as intended.

Original issue:

---
## Weasyprint Does Not Handle Multiple Font-Families in CSS for `<pre>` Tags

Please see the following StackOverflow question:

 http://stackoverflow.com/questions/21271711/how-do-i-specify-fallback-to-monospace-in-weasyprint

I have become convinced that this is actually an issue in the way that weasyprint handles the specification of multiple font-family values in CSS formatting for text in &lt;pre> tags.

**Comment by takis** [comment]
Hi, I think I have the same issue using:
- WeasyPrint 0.24
- OSX 10.10.5
- font-config 2.11.1
- pango 1.36.8
- cairo 1.14.2
- gdk-pixbuf 2.30.8

```
<!DOCTYPE html>
<html>
<body>
<p style="font: 32pt 'Comic Sans MS'">paragraph</p>
<p style="font: 32pt 'Comic Sans MS', sans-serif">paragraph</p>
<p style="font: 32pt sans-serif">paragraph</p>
</body>
</html>
```

When viewing this file in Chrome the first two paragraphs appear in Comic Sans as expected.

<img width="208" alt="ic-20150901-weasyprint_font_problem_chrome" src="https://cloud.githubusercontent.com/assets/128511/9609186/981abb00-50d2-11e5-9f7b-4cec09bb2b27.png">

When using WeasyPrint only the first paragraph appears in Comic Sans.
<img width="247" alt="ic-20150901-weasyprint_font_problem_weasyprint" src="https://cloud.githubusercontent.com/assets/128511/9609187/981e0210-50d2-11e5-842c-63123a20ab43.png">

**Comment by SimonSapin** [comment]
Weasyprint just [passes the `font-family` property through to `pango_font_description_set_family` as a comma-separated string of names](https://github.com/Kozea/WeasyPrint/blob/f0da0374bf0d54deb802393f6c636adf7541f9b9/weasyprint/text.py#L307-L308). If anyone figures out how to make Pango do the right thing, I’m interested.

**Comment by BodhiSukha** [comment]
I just realised I'm having a similar issue as well. I don't have an answer and but here is what I found so far. 

As far as I can understand, when provided with FAMILY_LIST pango's selection criteria isn't like what we expect. Here it says:
>For each of the resulting list of fonts, retrieve the coverage map for the current language tag, and then  **select the font which has the best PangoCoverageLevel** for the character. Ties are broken by ordering in the family list. Source: http://people.redhat.com/otaylor/pango-mirror/font-selection.shtml

Though even this doesn't make sense for my test case, because whenever I supply multiple fonts for fallback, same default font wins, which isn't in the list. 

I'm also unable to use comma separated family names in pango-view like this ```pango-view --no-display --output out.png --text '0123456789 - @ The lazy fox something something' --font "Open Sans, Times"``` . Getting the same results as weasyprint, which is that default font. 

So far, I just make sure I installed the font globally and provide only one font-family that I need.

**Comment by liZe** [comment]
I've tried hard to reproduce this bug, but I can't. `pango-view` works "as expected", and your example works very well for me under Linux. I use Pango 1.38.1.

As far as I can tell, it doesn't pick the font with the best language coverage. It just picks the first fonts it finds, for each character, as it generally does in a browser.

@SimonSapin it even looks really like CSS now, much more than it was in my memories. Any chance `pango_font_description_set_family` got better with time?

**Comment by SimonSapin** [comment]
¯\_(ツ)_/¯

**Comment by pzb** [comment]
I've run into the same issue and tracked it down to a little more detail.  It only happens on OS X.  `pango-view` demonstrates the error quite nicely.

`pango-view --no-display --output out1.png --text '0123456789 - @ The lazy fox something something' --font "Anonymous Pro Minus, monospace"` renders an image using a proportional sans-serif font (not monospace).
`pango-view --no-display --output out2.png --text '0123456789 - @ The lazy fox something something' --font "Anonymous Pro Minus"` renders using Anonymous.
`pango-view --no-display --output out3.png --text '0123456789 - @ The lazy fox something something' --font "monospace"` renders using a monospace font that is not Anonymous.
`pango-view --no-display --output out4.png --text '0123456789 - @ The lazy fox something something' --font "QuuxXyzzy"` renders identically to the first one.  QuuxXyzzy is a made up string that does not match any font on my system.

So it seems that listing multiple fonts on OS X is simply broken.  It does not choose the "best" one.  I'm guessing that is fails to split and treats it all as one font name, fails to match, and falls back to the default font.

**Comment by liZe** [comment]
@pzb thanks a lot for your debug information.

Looks like it's a bug in Pango. Fallback fonts are at least supported on MacOSX since [this old bug](https://bugzilla.gnome.org/show_bug.cgi?id=647969) has been fixed, at least in 1.32 (~summer 2012), is your version of Pango older than this? If not, could you report the bug on the [Pango bugzilla](https://bugzilla.gnome.org/)?

**Comment by liZe** [comment]
@pzb did you try a with recent version of Pango?

**Comment by pzb** [comment]
Yes.  https://bugzilla.gnome.org/show_bug.cgi?id=762873 is tracking the issue in pango.

**Comment by mb21** [comment]
> Weasyprint just passes the font-family property through

Would be great if this were mentioned in [the docs](http://weasyprint.readthedocs.io/en/latest/features.html#fonts), with an example like `font-family: Helvetica, sans-serif;` doesn't work, but `font-family: Helvetica;` does.

**Comment by liZe** [comment]
> with an example like font-family: Helvetica, sans-serif; doesn't work

@mb21 Actually, it works for everybody except OSX users because of a bug in an external library. I don't like the idea of adding a line about that in the "features" part of the documentation.

Documenting each bug for each version of each external library for each OS is an endless tiring job. Maybe adding a "Troubleshooting" section could help. I'll be happy to merge the pull requests :wink:.

**Comment by liZe** [comment]
It's been [fixed in Pango 1.42.3](https://gitlab.gnome.org/GNOME/pango/merge_requests/4).
